File: Lab_1/main_proc.py
import math
import sys

def check_root(ind):
    """Проверка коэфов -> возвращает один коэф"""
    try:
        global coef
        coef = sys.argv[ind]
    except:
        print("Введите коэффициент {} :".format(ind) )
        coef = input()
        #Проверка А на ноль(иначе впоследствие ошибка в делении на ноль)
        if float(coef) == 0.0 and ind == 1:
            print("Коэффициент 1 равен 0. Так не пойдет")
            check_root(1)
    return float(coef)

def get_roots(coef_list):
    """Получение значений коэфов -> возвращает список коэфов"""
    for i in range(0,3):
        coef_list[i] = check_root(i+1)
    return coef_list

def calculation(coef_list):
    """Подсчет корней -> возвращает множество корней"""
    A = coef_list[0]
    B = coef_list[1]
    C = coef_list[2]
    #Дискриминант
    D = B**2 - 4*A*C
    #Корни
    root_list = set()
    if (D>=0):
        root_list.add( (-B + math.sqrt(D))/ (2*A) )
        root_list.add( (-B - math.sqrt(D))/ (2*A) )
    return root_list

def print_ans(root_list):
    """Вывод ответов"""
    if len(root_list) == 0:
        print("Нет корней, дискриминант меньше нуля :(")
        return
    print("Корни:")
    for e in root_list:
        print(e, " ")

def all_process(coef_list):
    """Запускает последовательно все функции"""
    print_ans(calculation(get_roots(coef_list)))

def main():
    """Основная функция"""
    coef_list = [0.0]*3
    all_process(coef_list)

if __name__ == "__main__":
    main()

File: Lab_1-2-go/main_proc.go
package main

import (
	"bufio"   // для буферизованного ввода/вывода
	"fmt"     // для форматированного ввода/вывода
	"math"    // для математических функций
	"os"      // для взаимодействия с операционной системой
	"strconv" // для преобразования строк
	"strings" // для работы со строками
)

func main() {
	coef_list := get_roots() // Получаем коэффициенты
	all_process(coef_list)   // Обрабатываем коэффициенты
}

func all_process(coef_list []float64) {
	print_ans(calculation(coef_list)) // Вычисляем корни и выводим их
}

func print_ans(root_list []float64) {
	// Если корней нет, выводим сообщение
	if len(root_list) == 0 {
		fmt.Println("Нет корней, дискриминант меньше нуля :(")
	} else {
		fmt.Println("Корни:")
		// Выводим каждый корень
		for _, e := range root_list {
			fmt.Println(e, " ")
		}
	}
}

func calculation(coef_list []float64) []float64 {
	A := coef_list[0]
	B := coef_list[1]
	C := coef_list[2]
	D := math.Pow(B, 2) - 4*A*C // Вычисляем дискриминант
	root_list := make([]float64, 0)
	all_roots := make([]float64, 0)
	// Вычисляем корни, если дискриминант неотрицательный
	if D >= 0.0 {
		root_list = append(root_list, (-B+math.Sqrt(D))/(2.0*A))
		root_list = append(root_list, (-B-math.Sqrt(D))/(2.0*A))
	}
	// Добавляем корни в список
	for _, r := range root_list {
		if r >= 0 {
			all_roots = append(all_roots, math.Sqrt(r))
			all_roots = append(all_roots, -math.Sqrt(r))
		}
	}
	// Выводим коэффициенты и дискриминант
	for _, e := range coef_list {
		fmt.Println(e, " ", D)
	}
	return all_roots // Возвращаем список корней
}

func get_roots() []float64 {
	coef_list := make([]float64, 3) // Создаем список для коэффициентов
	for i := 0; i < 3; i++ {
		coef_list[i] = check_root(i + 1) // Проверяем каждый коэффициент
	}
	return coef_list
}

func check_root(ind int) float64 {
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Printf("Введите коэффициент %d: ", ind)
		input, _ := reader.ReadString('\n')
		coef, err := strconv.ParseFloat(strings.TrimSpace(input), 64)
		// Если введенное значение не является числом, выводим сообщение об ошибке
		if err != nil {
			fmt.Println("Ошибка. Попробуйте еще раз")
			continue
		}
		// Если первый коэффициент равен нулю, выводим сообщение об ошибке
		if coef == 0.0 && ind == 1 {
			fmt.Println("Коэффициент 1 равен 0. Так не пойдет")
			continue
		}
		return coef
	}
}
